yb-voyager/src/tgtdb/yugabytedb.go

```diff
@@ -300,10 +300,46 @@ var NonRetryCopyErrors = []string{
 	SYNTAX_ERROR,
 }
 
+// isDataIntegrityOrConstraintError checks if an error is a data integrity or constraint violation or syntax error
+// by examining the SQLSTATE code.
+//
+// SQLSTATE Class 22: Data Exception (e.g., 22003=numeric overflow, 22P02=invalid syntax)
+// SQLSTATE Class 23: Integrity Constraint Violation (e.g., 23502=not null, 23505=unique)
+// Postgres Error Codes: https://www.postgresql.org/docs/current/errcodes-appendix.html
+func isDataIntegrityOrConstraintError(err error) bool {
+	if err == nil {
+		return false
+	}
+
+	// Check pgx v4 errors
+	var pgErr *pgconn.PgError
+	if errors.As(err, &pgErr) {
+		code := pgErr.Code
+
+		if strings.HasPrefix(code, "22") || strings.HasPrefix(code, "23") {
```

Comment:
do you think we should add 42 as well? IT says "syntax error or access rule violation"? That sounds like a non-retryable error to me. Can you check that so that we don't have to come back to this problem again in the future?